### PR TITLE
Node.js Compat: Implement the nodejs_compat_v2 compat flag

### PR DESCRIPTION
**Pull request by jasnell**
The `nodejs_compat_v2` flag supersedes the 'nodejs_compat' flag. When enabled...

1. Node.js built-ins are available for import/require
2. Unlike the original `nodejs_compat` flag, Node.js imports do not require the 'node:' specifier prefix. Internally, the implementation will detect a Node.js raw specifier and convert it into the appropriate prefixed specifier, e.g. 'fs' becomes 'node:fs'
3. The `Buffer`, `process`, and `global` are exposed on the global
4. A user worker bundle can still provide its own implementations of all `node:` modules which will take precedence over the built-ins
5. The new `process.getBuiltinModule(...)` API is implemented. See https://github.com/nodejs/node/pull/52762

A worker can replace the implementation of `node:process` if they choose, which may mean that `getBuiltinModule(...)` is not available.

Refs: https://github.com/cloudflare/workerd/issues/2133

### CODE REVIEW

**Comment by IgorMinar** [comment]
@jasnell could you also expose `global` on `globalThis` as an alias of `globalThis`?

Essentially:
```js
globalThis.global = globalThis;
```

**Review by IgorMinar — APPROVED** [review]
I can't judge the implementation soundness, but I was able to play with this and can confirm that with the `nodejs_compat_v2` flag `Buffer` and `process` are now exposed, however `global` is not. Could you please add that?

I was also able to do some basic experiments with `process.getBuiltinModule` which also seems to work as expected. I did not test if overwriting a built-in using a user-provided module results in no change for the getBuiltinModule - this would be a good thing to add to an automated test.

Can you please add a test or sample for this functionality?

Otherwise this change looks good to me and will unblock our work on hybrid polyfills (https://github.com/unjs/unenv/issues/181). Thank you for a speedy turnaround.

To recap the todos:
- [ ] expose `global`
- [ ] add tests

FYI here is my test code:
```js
import path from 'path';
import npath from 'node:path';

export default {
  async fetch(req, env) {
    return new Response("Hello World\n" +
        (('Buffer' in globalThis) ? Buffer : "xBufferx") + '\n' +
        (('process' in globalThis) ? JSON.stringify(Object.keys(process)) : 'xprocessx') + '\n' +
        (JSON.stringify(Object.keys(process.getBuiltinModule('path')))) + '\n' +
        (JSON.stringify(Object.keys(path))) + '\n' +
        (JSON.stringify(Object.keys(npath))) + '\n'+
        (('global' in globalThis) ? global : "xglobalx") + '\n'
    );
  }
};
```

**Comment by jasnell** [comment]
The `getBuiltinModule(...)` implementation is not actually working correctly yet as verified by a test. Going to require just a bit more work to get right.

**Comment by jasnell** [comment]
@fhanau @harrishancock @dom96 ... this change is needed by the frameworks/wrangler folks to move forward on evolving the node.js compat story.

@irvinebroque @mikenovitch ... heads up from the product perspective this changes the way node.js compat is enabled. There's the existing `nodejs_compat` flag for the original version, and then a new `nodejs_compat_v2` flag to enable this variation. Whee!

**Comment by IgorMinar** [comment]
> @irvinebroque @mikenomitch ... heads up from the product perspective this changes the way node.js compat is enabled. There's the existing nodejs_compat flag for the original version, and then a new nodejs_compat_v2 flag to enable this variation. Whee!

Is the plan to make `nodejs_compat_v2` become equivalent with `nodejs_compat` after a certain compat date? That would be my preference as I want to avoid a proliferation of compat flags. I see `nodejs_compat_v2` just as a temporary vehicle to test the changes we plan to make to `nodejs_compat` which will be enabled at a set compat date.

File: src/workerd/api/node/node-compat-v2-test.wd-test [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  services = [
+    ( name = "node-compat-v2-test",
+      worker = (
+        modules = [
+          (name = "worker", esModule = embed "node-compat-v2-test.js"),
+          (name = "node:fs", esModule = "export default 1"),
+          (name = "node:path", esModule = "export default 2"),
+        ],
+        compatibilityDate = "2024-05-01",
+        compatibilityFlags = ["nodejs_compat_v2", "experimental"]
```
Comment (IgorMinar) [inline src/workerd/api/node/node-compat-v2-test.wd-test:13]: is it sufficient to just add `"experimental"` into copatiblityFlags array? I tried this a modified version of one of the existing samples and got the following:

```
bazel-bin/src/workerd/server/workerd serve ./samples/helloworld_esm/config.capnp 
service main: The compatibility flag experimental is experimental and may break or be removed in a future version of workerd. To use this flag, you must pass --experimental on the command line.
```

only `bazel-bin/src/workerd/server/workerd serve ./samples/helloworld_esm/config.capnp --experimental` worked for me.

Comment (IgorMinar) [reply src/workerd/api/node/node-compat-v2-test.wd-test:13]: my config.capnp looked like this:
```
const helloWorld :Workerd.Worker = (
  modules = [
    (name = "worker", esModule = embed "worker.js")
  ],
  compatibilityDate = "2024-05-12",
  compatibilityFlags = ["nodejs_compat", "nodejs_compat_v2", "experimental"]
);

```

**Comment by jasnell** [comment]
> Is the plan to make nodejs_compat_v2 become equivalent with nodejs_compat after a certain compat date? That would be my preference as I want to avoid a proliferation of compat flags. I see nodejs_compat_v2 just as a temporary vehicle to test the changes we plan to make to nodejs_compat which will be enabled at a set compat date.

Unfortunately I don't think so. Given the way compat dates work and the guarantees they carry, we can't really combine these at any point and we had made the decision long ago that Node.js compat would not be on-by-default ever. If we move forward with adding this, it'll mean a second node.js compat mode going forward indefinitely. This is something we should discuss tho to see if there's a reasonable path forward.

File: src/workerd/api/node/node-compat-v2-test.wd-test (hunk 1 above)
Comment (jasnell) [reply src/workerd/api/node/node-compat-v2-test.wd-test:13]: The `--experimental` label is there only while this is being evaluated. I plan to remove it before landing if we do decide to move forward on this. I'd like to make sure we get input from @irvinebroque and @mikenomitch

**Comment by IgorMinar** [comment]
@jasnell sorry if I was not clear, I'm not suggesting that `nodejs_compat_v2` becomes activated on by default after some compat date. What I'm suggesting instead is that after some compat date `nodejs_compat` and `nodejs_compat_v2` mean the same thing (and `nodejs_compat_v2` becomes unnecessary). This would reduce the number of flags we need to document and explain to developers as well as number of branches in our codebase.

File: src/workerd/api/node/node-compat-v2-test.wd-test (hunk 1 above)
Comment (IgorMinar) [reply src/workerd/api/node/node-compat-v2-test.wd-test:13]: I think I didn't make myself clear, what I'm puzzled about is that you are able to enable the experimental mode via a runtime flag in the config, but for me this didn't work and I had to explicitly use `--experimental` command line flag 
instead. Is that expected?

I don't know how this stuff works, and I tried doing what you did here at first but it didn't work, so when I later saw you using this approach in this PR I was surprised.

**Comment by jasnell** [comment]
Even then, I'm not sure we can combine `nodejs_compat` and `nodejs_compat_v2` unfortunately. This PR introduces a few changes that could be technically breaking even in the `nodejs_compat` case currently. We'll need to evaluate carefully.

**Comment by jasnell** [comment]
> ... is the main breaking change the new globals that are added?

I'll say "potentially" breaking change are the new globals and the fact that statements like `import('buffer')` will suddenly start working where they fail now. As we've seen with other recent changes, folks like to do feature-sniffing by trying an action and revering to some fall back logic when it doesn't work. Having it so that the non-prefixed specifiers suddenly start working could cause unexpected side effects that we need to be careful of.

File: src/workerd/io/compatibility-date.capnp [hunk 2]
```diff
@@ -441,4 +441,13 @@ struct CompatibilityFlags @0x8f8c1b68151b6cef {
...
   # Ensures that WHATWG standard URL parsing is used in the fetch API implementation.
+
+  nodeJsCompatV2 @50 :Bool
+      $compatEnableFlag("nodejs_compat_v2")
+      $compatDisableFlag("no_nodejs_compat_v2")
+      $experimental;
+  # Implies nodeJSCompat with the following additional modifications:
+  # * Node.js Compat built-ins may be imported/required with or without the node: prefix
+  # * Node.js Compat common globals are available everywhere (e.g. Buffer, process)
+  # * node:process cannot be replaced by the worker bundle
```
Comment (fhanau) [inline src/workerd/io/compatibility-date.capnp:452]: Where is this implemented?

File: src/workerd/api/node/node-compat-v2-test.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    // But process.getBuiltinModule should always return the built-in module.
+    const builtInPath = process.getBuiltinModule('node:path');
+    const builtInFs = process.getBuiltinModule('node:fs');
+
+    // These are from the worker bundle....
+    assert.strictEqual(fs, 1);
+    assert.strictEqual(path, 2);
+
+    // But these are from the built-ins...
+    // node:fs is not implementated currently so it should be undefined here.
```
Comment (fhanau) [inline src/workerd/api/node/node-compat-v2-test.js:47]: nit: spelling

File: src/node/internal/process.ts [hunk 4]
```diff
@@ -62,7 +64,12 @@ export const env = new Proxy({}, {
   }
 });
 
+export function getBuiltinModule(id : string) : any {
```
Comment (fhanau) [inline src/node/internal/process.ts:67]: nit: fix spaces

File: src/workerd/io/compatibility-date.capnp [hunk 5]
```diff
@@ -441,4 +441,13 @@ struct CompatibilityFlags @0x8f8c1b68151b6cef {
     $compatDisableFlag("fetch_legacy_url")
     $compatEnableDate("2024-06-03");
   # Ensures that WHATWG standard URL parsing is used in the fetch API implementation.
+
+  nodeJsCompatV2 @50 :Bool
```
Comment (fhanau) [inline src/workerd/io/compatibility-date.capnp:445]: For visibility, should we place this next to the existing `nodeJsCompat` flag?

File: src/workerd/api/global-scope.c++ [hunk 6]
```diff
@@ -777,6 +777,48 @@ void ServiceWorkerGlobalScope::reportError(jsg::Lock& js, jsg::JsValue error) {
   }
 }
 
+namespace {
+jsg::JsValue resolveFromRegistry(jsg::Lock& js, kj::StringPtr specifier) {
+  auto moduleRegistry = jsg::ModuleRegistry::from(js);
+  if (moduleRegistry == nullptr ) {
+    // TODO: Return the known built-in instead?
```
Comment (fhanau) [inline src/workerd/api/global-scope.c++:784]: Decide how to approach this TODO, there's also minor lint in the line above.

**Comment by fhanau** [comment]
LGTM otherwise

File: src/workerd/io/compatibility-date.capnp (hunk 2 above)
Comment (jasnell) [reply src/workerd/io/compatibility-date.capnp:452]: oh right... decided not to go with this. Just need to remove this bullet point

File: src/workerd/io/compatibility-date.capnp (hunk 5 above)
Comment (jasnell) [reply src/workerd/io/compatibility-date.capnp:445]: I'd rather keep the ordinal values in order to make it easier to know which is the next available value.

File: src/node/internal/process.ts (hunk 4 above)
Comment (jasnell) [reply src/node/internal/process.ts:67]: For stuff like this... using the "Add a suggestion" option in the github editor is easier.

File: src/workerd/api/global-scope.c++ [hunk 7]
```diff
@@ -777,6 +777,48 @@ void ServiceWorkerGlobalScope::reportError(jsg::Lock& js, jsg::JsValue error) {
   }
 }
 
+namespace {
+jsg::JsValue resolveFromRegistry(jsg::Lock& js, kj::StringPtr specifier) {
+  auto moduleRegistry = jsg::ModuleRegistry::from(js);
+  if (moduleRegistry == nullptr ) {
```
Comment (dom96) [inline src/workerd/api/global-scope.c++]:
```suggestion
  if (moduleRegistry == nullptr) {
```

File: src/workerd/jsg/modules.h [hunk 8]
```diff
@@ -441,12 +444,8 @@ class ModuleRegistryImpl final: public ModuleRegistry {
 
   void addBuiltinModule(Module::Reader module) {
     if (module.which() != Module::SRC) {
-      using Key = typename Entry::Key;
       auto specifier = module.getName();
       auto path = kj::Path::parse(specifier);
-      if (module.getType() == Type::BUILTIN && entries.find(Key(path, Type::BUNDLE)) != kj::none) {
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/jsg/modules.h:447]: Previously, we would skip registering the built-in module at all if the worker bundle itself included module with the same name. That was really just an optional optimization. With the introduction of `process.getBuiltinModule(...)`, we want the built-ins to be registered and accessible even if the worker bundle shadows it with its own implementation.

File: src/workerd/api/node/node-compat-v2-test.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    const { default: process } = await import('node:process');
+    assert.strictEqual(process, globalThis.process);
+
+    assert.strictEqual(global, globalThis);
+  }
+};
+
+export const nodeJsGetBuiltins = {
+  async test() {
+    // node:* modules in the worker bundle should override the built-in modules...
```
Comment (dom96) [inline src/workerd/api/node/node-compat-v2-test.js:32]: Does the same apply to non-`node:` prefixed modules? Do we want tests for it too?

File: src/workerd/io/compatibility-date.capnp [hunk 10]
```diff
@@ -441,4 +441,12 @@ struct CompatibilityFlags @0x8f8c1b68151b6cef {
...
+      $compatEnableFlag("nodejs_compat_v2")
+      $compatDisableFlag("no_nodejs_compat_v2")
+      $experimental;
+  # Implies nodeJSCompat with the following additional modifications:
+  # * Node.js Compat built-ins may be imported/required with or without the node: prefix
+  # * Node.js Compat the globals Buffer and process are available everywhere
```
Comment (dom96) [inline src/workerd/io/compatibility-date.capnp:451]:
```suggestion
  # * The Node.js globals Buffer and process are available everywhere
```

File: src/workerd/io/worker.c++ [hunk 11]
```diff
@@ -1018,6 +1018,9 @@ Worker::Isolate::Isolate(kj::Own<Api> apiParam,
 
     lock->setCaptureThrowsAsRejections(features.getCaptureThrowsAsRejections());
     lock->setCommonJsExportDefault(features.getExportCommonJsDefaultNamespace());
+    if (features.getNodeJsCompatV2()) {
+      lock->setNodeJsCompatEnabled();
```
Comment (dom96) [inline src/workerd/io/worker.c++:1022]: Should `setNodeJsCompatEnabled` mention the "v2"?

File: src/workerd/jsg/jsg.h [hunk 12]
```diff
@@ -2296,6 +2296,8 @@ class Lock {
   void setCaptureThrowsAsRejections(bool capture);
   void setCommonJsExportDefault(bool exportDefault);
 
+  void setNodeJsCompatEnabled();
```
Comment (dom96) [inline src/workerd/jsg/jsg.h:2299]: A comment on what this enables (v1. vs. v2) would be nice here

File: src/workerd/io/compatibility-date.capnp (hunk 10 above)
Comment (dom96) [reply src/workerd/io/compatibility-date.capnp:451]: Also, why is `Buffer` and `process` only made available? Are we planning to make more of the builtins available too?

**Review by dom96 — APPROVED** [review]

**Review by fhanau — APPROVED** [review]

File: src/workerd/api/node/node-compat-v2-test.js (hunk 9 above)
Comment (jasnell) [reply src/workerd/api/node/node-compat-v2-test.js:32]: It does but tests for that should be added as a separate PR.

File: src/workerd/io/compatibility-date.capnp (hunk 10 above)
Comment (jasnell) [reply src/workerd/io/compatibility-date.capnp:451]: These are the only Node.js specific global properties that we implement. Most of the remaining globals supported by Node.js are either web platform standard APIs we already implement/export, are ones we don't implement (e.g. `BroadcastChannel`), or that don't make sense in our environment, e.g. `__filename` and `__dirname`.

File: src/workerd/io/worker.c++ (hunk 11 above)
Comment (jasnell) [reply src/workerd/io/worker.c++:1022]: No, eventually the goal is to combine the two compat flags if we can. In either case, there's no reason to version this particular API.

File: src/workerd/jsg/jsg.h (hunk 12 above)
Comment (jasnell) [reply src/workerd/jsg/jsg.h:2299]: At this level it doesn't matter as jsg really has (and should not have) any information directly relating to compatibility flags. This is just a JSG specific configuration option that the compat flag happens to enable. We don't want any specific versioning on this.
